controller/DeliveryServer.php

```diff
@@ -65,56 +63,10 @@ public function finishDeliveryExecution()
         if ($deliveryExecution->getState()->getUri() == ProctoredDeliveryExecution::STATE_PAUSED) {
             $redirectUrl = _url('awaitingAuthorization', 'DeliveryServer', 'ltiProctoring', ['deliveryExecution' => $deliveryExecution->getIdentifier()]);
         } else {
-            $redirectUrl = LTIDeliveryTool::singleton()->getFinishUrl($this->getLtiMessage($deliveryExecution), $deliveryExecution);
-        }
-        $this->redirect($redirectUrl);
-    }
-
-    /**
-     * @param DeliveryExecution $deliveryExecution
-     * @return LtiMessage
-     */
-    protected function getLtiMessage(DeliveryExecution $deliveryExecution)
-    {
-        $state = $deliveryExecution->getState()->getLabel();
-        /** @var DeliveryLog $deliveryLog */
-        $deliveryLog = $this->getServiceManager()->get(DeliveryLog::SERVICE_ID);
-        $reason = '';
-        $reasons = null;
-        switch ($deliveryExecution->getState()->getUri()) {
-            case ProctoredDeliveryExecution::STATE_FINISHED:
-                $log = $deliveryLog->get($deliveryExecution->getIdentifier(), 'TEST_EXIT_CODE');
-                if ($log) {
-                    $reason .= 'Exit code: ' . $log[count($log) - 1]['data']['exitCode'] . PHP_EOL;
-                }
-                break;
-            case ProctoredDeliveryExecution::STATE_TERMINATED:
-                $log = $deliveryLog->get($deliveryExecution->getIdentifier(), 'TEST_TERMINATE');
-                if ($log) {
-                    $reasons = $log[count($log) - 1]['data'];
-                }
-                break;
-            case ProctoredDeliveryExecution::STATE_PAUSED:
-                $log = $deliveryLog->get($deliveryExecution->getIdentifier(), 'TEST_PAUSE');
-                if ($log) {
-                    $reasons = $log[count($log) - 1]['data'];
-                }
-                break;
-            case ProctoredDeliveryExecution::STATE_CANCELED:
-                $log = $deliveryLog->get($deliveryExecution->getIdentifier(), 'TEST_CANCEL');
-                if ($log) {
-                    $reasons = $log[count($log) - 1]['data'];
-                }
-                break;
+            $redirectUrl = LTIDeliveryTool::singleton()->getFinishUrl($deliveryExecution);
```

Comment:
That is not good to use this singleton, probably we could try to replace it with service?

Reply:
As I answered [here](https://github.com/oat-sa/extension-tao-ltideliveryprovider/pull/249#discussion_r425138118) it's not a new code and I think it's better to refactor it in a separate PR since it will be a bigger change.

Reply:
I wrote that before your answer